### PR TITLE
RS-100: blog how to implement async storage data streaming in your py torch project

### PR DESCRIPTION
**Pull request by AnthonyCvn**
New blog post about "How to Implement Async Storage Data Streaming in Your PyTorch Project"

### CODE REVIEW

File: blog/2024-01-12-implement-async-storage-data-streaming-pytorch/index.md [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+---
+
+![PyTorch Training Diagram](./img/pytorch-iterabledataset.webp)
+<small>PyTorch training loop with data streaming from remote device</small>
+
+When training a model, we aim to process data in batches, shuffle data at each epoch to avoid over fitting, and leverage Python's `multiprocessing` for data fetching through multiple workers.
+
+The reason that we want to use multiple workers is that GPUs are capable of handling large amounts of data concurrently; however, the bottleneck often lies in the time-consuming task of loading this data into the system.
+
+Moreover, the challenge is even trickier when there is simply too much data to store the whole dataset on disk and we need to stream data from an async storage database such as [**ReductStore**](<https://www.reduct.store/>).
```
Comment (atimin) [inline blog/2024-01-12-implement-async-storage-data-streaming-pytorch/index.md:18]: async storage database - I would say data storage or just database. I'm also not sure if we should use async here. I use an asynchronous approach to develop ReductStore, but it is internal. We use the word async mostly when we talk about client SDKs.

File: blog/2024-01-12-implement-async-storage-data-streaming-pytorch/index.md [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+    <!-- -->
+
+- Initialize the client:
+
+
+<!-- -->
+
+```python
+client = Client(HOST, api_token=API_TOKEN)
```
Comment (atimin) [inline blog/2024-01-12-implement-async-storage-data-streaming-pytorch/index.md:53]: I suggest adding the steps here that you describe above. I would also show the snippet first, and then explain it.

File: blog/2024-01-12-implement-async-storage-data-streaming-pytorch/index.md [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+```
+
+### Getting data from playground server
+To use the 'Cats' dataset with Python, you'll need to follow a few other straightforward steps.
+
+First, ensure you have installed necessary libraries like NumPy and OpenCV.
+
+Then access your dataset by asynchronously querying records within a specified range using, e.g. `bucket.query(DATASET, start=0, stop=5)`.
```
Comment (atimin) [inline blog/2024-01-12-implement-async-storage-data-streaming-pytorch/index.md:61]: here I would explain that ReductStore is a time-series database and we usually use timestamps to specify the range. But the timestamps are a number of microseconds from 1970-01-01, so we can use them as IDs of records.

File: blog/2024-01-12-implement-async-storage-data-streaming-pytorch/index.md [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+Implementing a Buffered Shuffle Iterable Dataset in PyTorch is a trade-off to be able to shuffle the data we fetch from the remote server. This works, but we have to be careful not to get duplicates when using several workers, and we need to define a different data stream for each seperate process. 
+
+One way to handle this, is to look at the worker information for each separate process and request different data ranges from the database.
+
+```python
+import random
+import torch
+
+class BufferedShuffleIterableDataset(RemoteCatsDataset):
```
Comment (atimin) [inline blog/2024-01-12-implement-async-storage-data-streaming-pytorch/index.md:210]: This is really interesting. I could implement it on the storage side. A user could use something like this:

```python
bucket.query(self.dataset_name, start=worker_start, stop=worker_stop, random=true)
```

**Review by atimin — CHANGES_REQUESTED** [review]

File: blog/2024-01-12-implement-async-storage-data-streaming-pytorch/index.md (hunk 4 above)
Comment (AnthonyCvn) [reply blog/2024-01-12-implement-async-storage-data-streaming-pytorch/index.md:210]: That would be great! We should definitely keep that in backlog.

**Pull request merged**